File: 061.pl
#!/usr/bin/env perl
use v5.14;
use warnings;
use List::Util qw(sum);

my %func = (
    '3' => sub { $_[0] * ( $_[0] + 1 ) / 2 },
    '4' => sub { $_[0] * $_[0] },
    '5' => sub { $_[0] * ( 3 * $_[0] - 1 ) / 2 },
    '6' => sub { $_[0] * ( 2 * $_[0] - 1 ) },
    '7' => sub { $_[0] * ( 5 * $_[0] - 3 ) / 2 },
    '8' => sub { $_[0] * ( 3 * $_[0] - 2 ) }
);

my ( $n, $num, %nums, @type8 );
while (1) {
    $n++;
    foreach my $x ( reverse 3 .. 8 ) {
        $num = $func{$x}->($n);
        if ( $num ~~ [ 1000 .. 9999 ] ) {
            if ( $x == 8 ) {
                push @type8, $num;
            }
            else {
                push @{ $nums{ substr( $num, 0, 2 ) } }, [ $num, $x ];
            }
        }
    }
    last if $num > 9999;
}
solve( ['8'], [$_] ) foreach @type8;

sub solve {
    my ( $types, $num_set ) = @_;

    if ( @$types == 6 ) {
        if ( substr( $num_set->[0], 0, 2 ) eq substr( $num_set->[-1], -2 ) ) {
            say sum(@$num_set);
        }
    }
    else {
        foreach my $n ( @{ $nums{ substr( $num_set->[-1], -2 ) } } ) {
            unless ( $n->[1] ~~ @$types ) {
                solve( [ @$types, $n->[1] ], [ @$num_set, $n->[0] ] );
            }
        }
    }
}

File: 061.rb
#!/usr/bin/env ruby -wKU
func = {
  3 => ->(x) {x * (x + 1) / 2},
  4 => ->(x) {x * x},
  5 => ->(x) {x * (3 * x - 1) / 2},
  6 => ->(x) {x * (2 * x - 1)},
  7 => ->(x) {x * (5 * x - 3) / 2},
  8 => ->(x) {x * (3 * x - 2)}
}

n, num, $type8 = 0, 0, []
$nums = Hash.new { |hash, key| hash[key] = [] }
while true
  n += 1
  (3 .. 8).reverse_each do |x|
    num = func[x].call(n)
    if (1000 ... 10000).include?(num)
      if x == 8
        $type8 << num
      else
        $nums[num / 100] << [num, x]
      end
    end
  end
  break if num > 9999
end

def solve(num_set, types)
  if types.size == 6
    if num_set[0] / 100 == num_set[-1] % 100
      puts num_set.reduce(:+)
    end
  else
    $nums[num_set[-1] % 100].each do |n|
      unless types.include?(n[1])
        solve(num_set + [n[0]], types + [n[1]])
      end
    end
  end
end
$type8.each { |e| solve([e], [8]) }
